homeassistant/components/rainbird/__init__.py

```diff
@@ -36,6 +41,10 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
             entry.data[CONF_PASSWORD],
         )
     )
+
+    if entry.unique_id is None:
+        await _fix_unique_id(hass, controller, entry)
```

Comment:
Why do we not migrate the config entries with serial number as unique id?

Reply:
My thought was that not all existing serial numbers are invalid, so it would just avoid migrations that are optional. It leaves some complexity in the possible states that exist, but has less code complexity.

My impression is that we'd also need to keep this in sync with existing entities in the entity registry, so all of those would also need to be rewritten (and keep code around to manage this), is that right?

Reply:
Requested additional feedback. Let me know your thoughts and I can proceed with additional changes.

Reply:
If we don't migrate the existing unique ids with serial number the already configured check won't catch those devices if they are tried again.

Reply:
My impression is the existing config entry tests exercise this case and they shouldn't be reconfigured (though maybe there is a gap in missing, I can take a closer look)

Reply:
I think I realize that i may need the code to migrate entities to unique ids anyway. Some entities before this had an id of "0" and then will have a different unique id after this.

Reply:
Updated this to fix all config entries and all entity unique ids.